packages/cdk/resources/Cognito.ts

```diff
@@ -0,0 +1,207 @@
+
+  public constructor(scope: Construct, id: string, props: CognitoProps) {
+    super(scope, id)
+
+    // Resources
+    const userPool = new UserPool(this, "UserPool", {
+      removalPolicy: RemovalPolicy.DESTROY
+    })
+
+    const userPoolDomain = new UserPoolDomain(this, "UserPoolDomain", {
+      userPool,
+      customDomain: {
+        domainName: props.fullCognitoDomain,
+        certificate: props.cognitoCertificate
+      }
+    })
+
+    new ARecord(this, "UserPoolCloudFrontAliasIpv4Record", {
+      zone: props.hostedZone,
+      recordName: props.shortCognitoDomain,
+      target: RecordTarget.fromAlias(new UserPoolDomainTarget(userPoolDomain))
+    })
+
+    new AaaaRecord(this, "UserPoolCloudFrontAliasIpv6Record", {
+      zone: props.hostedZone,
+      recordName: props.shortCognitoDomain,
+      target: RecordTarget.fromAlias(new UserPoolDomainTarget(userPoolDomain))
+    })
+
+    // these are the endpoints that are added to user pool identity provider
+    // note we override the token endpoint to point back to our custom token
+    const oidcEndpoints: OidcEndpoints = {
+      authorization: props.primaryOidcAuthorizeEndpoint,
+      jwksUri: props.primaryOidcjwksEndpoint,
+      token: `https://${props.fullCloudfrontDomain}/api/token`,
+      userInfo: props.primaryOidcUserInfoEndpoint
+    }
+
+    // eslint-disable-next-line max-len
+    // see https://digital.nhs.uk/services/care-identity-service/applications-and-services/cis2-authentication/guidance-for-developers/detailed-guidance/scopes-and-claims
+    // about claims that we need
+    const primaryPoolIdentityProvider = new UserPoolIdentityProviderOidc(this, "UserPoolIdentityProvider", {
+      name: "Primary", // this name is used in the web client
+      clientId: props.primaryOidcClientId,
+      clientSecret: props.primaryOidClientSecret,
+      issuerUrl: props.primaryOidcIssuer,
+      userPool: userPool,
+      attributeRequestMethod: OidcAttributeRequestMethod.GET,
+      scopes: ["openid", "profile", "email", "nhsperson", "nationalrbacaccess"],
+      endpoints: oidcEndpoints
+    })
+
+    const supportedIdentityProviders: Array<UserPoolClientIdentityProvider> = [
+      UserPoolClientIdentityProvider.COGNITO,
+      UserPoolClientIdentityProvider.custom(primaryPoolIdentityProvider.providerName)
+    ]
+
+    // define some variables that we need for mock authorization
+    let mockPoolIdentityProvider!: UserPoolIdentityProviderOidc
+
+    // set up things for mock login
+    if (props.useMockOidc) {
+      if (props.mockOidcAuthorizeEndpoint === undefined ||
+        props.mockOidcjwksEndpoint === undefined ||
+        props.mockOidcUserInfoEndpoint === undefined ||
+        props.mockOidcClientId === undefined ||
+        props.mockOidClientSecret === undefined ||
+        props.mockOidcIssuer === undefined
+      ) {
+        throw new Error("Attempt to use mock oidc but variables are not defined")
+      }
```

Comment:
This seems a bit OTT, we're not doing the same checks for the primary props.

Reply:
this is needed for typing. 
The props in the if statement are optional to this construct and are only passed in when useMockOidc is true. 
If we dont have the if statement, then typescript thinks they may be undefined and so throws errors when compiling the resources inside the if statement